nipoppy_new/nipoppy/workflows/pipeline.py

```diff
@@ -0,0 +1,402 @@
+        bids_layout: bids.BIDSLayout = create_bids_db(
+            dpath_bids=self.layout.dpath_bids,
+            dpath_bids_db=dpath_bids_db,
+            ignore_patterns=pybids_ignore_patterns,
+            reset_database=True,
+        )
+
+        # list all the files in BIDSLayout
+        # since we are selecting for specific a specific subject and
+        # session, there should not be too many files
+        filenames = bids_layout.get(return_type="filename")
+        self.logger.debug(f"Found {len(filenames)} files in BIDS database:")
+        for filename in filenames:
+            self.logger.debug(filename)
+
+        if len(filenames) == 0:
+            self.logger.warning("BIDS database is empty")
+
+        return bids_layout
+
+    def check_dir(self, dpath: Path):
+        """Create directory if it does not exist."""
+        if not dpath.exists():
+            self.mkdir(dpath, log_level=logging.WARNING)
+
+    def run_setup(self, **kwargs):
+        """Run pipeline setup."""
+        to_return = super().run_setup(**kwargs)
+
+        # make sure the pipeline config exists
+        self.pipeline_config
+
+        for dpath in self.dpaths_to_check:
+            self.check_dir(dpath)
+
+        return to_return
+
+    def run_main(self, **kwargs):
+        """Run the pipeline."""
+        for participant, session in self.get_participants_sessions_to_run(
+            self.participant, self.session
+        ):
+            self.logger.info(f"Running on participant {participant}, session {session}")
+            try:
+                self.run_single(participant, session)
+            except Exception as exception:
+                self.logger.error(
+                    f"Error running {self.pipeline_name} {self.pipeline_version}"
+                    f" on participant {participant}, session {session}"
+                    f": {exception}"
+                )
+
+    def run_cleanup(self, **kwargs):
+        """Run pipeline cleanup."""
+        if self.dpath_pipeline_work.exists():
+            self.rm(self.dpath_pipeline_work)
+        return super().run_cleanup(**kwargs)
+
+    def get_participants_sessions_to_run(
+        self, participant: Optional[str], session: Optional[str]
+    ):
+        """Return participant-session pairs to run the pipeline on."""
+        # TODO add option in Boutiques descriptor of pipeline
+        # 1. "manifest" (or "all"?)
+        # 2. "downloaded" (from doughnut)
+        # 3. "organized" (from doughnut)
+        # 4. "converted" (from doughnut)
+        # 5. "dataset" (i.e. apply on entire dataset, do not loop over anything)
+
+        # for now just check the participants/sessions that have BIDS data
+        return self.doughnut.get_converted_participants_sessions(
+            participant=participant, session=session
+        )
+
+    @abstractmethod
+    def run_single(self, participant: Optional[str], session: Optional[str]):
```

Comment:
Not sure since `BaseWorkflow` does not necessarily do any looping of participant/sessions (e.g. in the `InitWorkflow` and the eventual "import BIDS dataset to Nipoppy" workflow)